File: 06/breadth-first_search_practice0602.py
from collections import deque

graph = {}
graph["CAB"] = ["CAT", "CAR"]
graph["CAT"] = ["MAT", "BAT", "CAR"]
graph["CAR"] = ["CAT", "BAR"]
graph["MAT"] = ["BAT"]
graph["BAR"] = ["BAT"]

def search(start):
    search_queue = deque()
    search_queue += graph[start]
    searched = []
    while search_queue:
        position = search_queue.popleft()

        if not position in searched:
            if position == "BAT":
                return True
            else:
                search_queue += graph[position]
                searched.append(position)
    
    return False

print(search("CAB"))

File: 06/breadthFirstSearchPractice0602.js
const graph = {};

graph["CAB"] = ["CAT", "CAR"];
graph["CAT"] = ["MAT", "BAT", "CAR"];
graph["CAR"] = ["CAT", "BAR"];
graph["MAT"] = ["BAT"];
graph["BAR"] = ["BAT"];

const search = (start) => {
	let search_queue = [];
	search_queue = [...search_queue, ...graph[start]];
	let searched = [];

	while (search_queue.length !== 0) {
		const position = search_queue.shift();
		if (!searched.includes(position)) {
			if (position.localeCompare("BAT")) {
				return true;
			} else {
				search_queue = [...search_queue, graph[position]];
				searched = [...searched, position];
			}
		}
	}

	return false;
};

console.log(search("CAB"));
